modules/sor/sorV2/lib/pools/stable/stablePool.ts

```diff
@@ -0,0 +1,364 @@
+        amp: bigint,
+        swapFee: bigint,
+        tokens: StablePoolToken[],
+        totalShares: bigint,
+        tokenPairs: TokenPairData[],
+    ) {
+        this.chain = chain;
+        this.id = id;
+        this.address = address;
+        this.amp = amp;
+        this.swapFee = swapFee;
+        this.totalShares = totalShares;
+
+        this.tokens = tokens.sort((a, b) => a.index - b.index);
+        this.tokenMap = new Map(this.tokens.map((token) => [token.token.address, token]));
+        this.tokenIndexMap = new Map(this.tokens.map((token) => [token.token.address, token.index]));
+        this.tokenPairs = tokenPairs;
+    }
+
+    addLiquiditySingleTokenExactIn(
+        tokenIn: Token,
+        bpt: Token,
+        amount: TokenAmount,
+        mutateBalances?: boolean,
+    ): TokenAmount {
+        try {
+            // balances and amounts must be normalized to 1e18 fixed point - e.g. 1USDC => 1e18 not 1e6
+            const tokenBalances: bigint[] = [];
+            const amountsIn: bigint[] = [];
+            Array.from(this.tokenMap.values()).forEach((stablePoolToken) => {
+                if (stablePoolToken.token.isSameAddress(tokenIn.address)) {
+                    amountsIn.push(amount.scale18);
+                } else {
+                    amountsIn.push(0n);
+                }
+                tokenBalances.push(stablePoolToken.scale18);
+            });
+            const currentInvariant = _calculateInvariant(this.amp, tokenBalances);
+            const bptAmountOut = _calcBptOutGivenExactTokensIn(
+                this.amp,
+                tokenBalances,
+                amountsIn,
+                this.totalShares,
+                currentInvariant,
+                this.swapFee,
+            );
+            return TokenAmount.fromRawAmount(bpt, bptAmountOut);
+        } catch (err) {
+            return TokenAmount.fromRawAmount(bpt, 0n);
+        }
+    }
+
+    addLiquiditySingleTokenExactOut(
+        tokenIn: Token,
+        bpt: Token,
+        amount: TokenAmount,
+        mutateBalances?: boolean,
+    ): TokenAmount {
+        try {
+            const tokenInIndex = this.tokenIndexMap.get(tokenIn.address);
+            if (!tokenInIndex) throw new Error('Provided Token In is Invalid');
+            const tokenBalances = Array.from(this.tokens.values()).map(({ scale18 }) => scale18);
+            const currentInvariant = _calculateInvariant(this.amp, tokenBalances);
+            const tokenInAmount = _calcTokenInGivenExactBptOut(
+                this.amp,
+                tokenBalances,
+                tokenInIndex,
+                amount.scale18,
+                this.totalShares,
+                currentInvariant,
+                this.swapFee,
+            );
+            return TokenAmount.fromRawAmount(tokenIn, tokenInAmount);
+        } catch (err) {
+            return TokenAmount.fromRawAmount(tokenIn, 0n);
+        }
+    }
+
+    getLimitAmountAddLiquidity(tokenIn: Token): bigint {
```

Comment:
It feels weird to have a limit for add liquidity amounts 🤔 
Where did you get this from?